controllers/default_organization_controller.go

```diff
@@ -0,0 +1,87 @@
+package controllers
+
+import (
+	"context"
+
+	"go.uber.org/multierr"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	controlv1 "github.com/appuio/control-api/apis/v1"
+)
+
+// DefaultOrganizationReconciler reconciles User resources to ensure they have a DefaultOrganization set if applicable.
+type DefaultOrganizationReconciler struct {
+	client.Client
+	Recorder record.EventRecorder
+	Scheme   *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=appuio.io,resources=organizationmembers,verbs=get;list;watch
+//+kubebuilder:rbac:groups=appuio.io,resources=users,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=appuio.io,resources=users/status,verbs=get
+
+// Reconcile reacts on changes of users and mirrors these changes to Keycloak
+func (r *DefaultOrganizationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.V(1).WithValues("request", req).Info("Reconciling")
+
+	memb := controlv1.OrganizationMembers{}
+	if err := r.Get(ctx, req.NamespacedName, &memb); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !memb.ObjectMeta.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	allMemberships := controlv1.OrganizationMembersList{}
+	if err := r.List(ctx, &allMemberships); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var errGroup error
+	for _, user := range memb.Status.ResolvedUserRefs {
+		myOrgs := make([]string, 0)
+
+		for _, membership := range allMemberships.Items {
+			for _, membershipUser := range membership.Status.ResolvedUserRefs {
+				if user.Name == membershipUser.Name {
+					myOrgs = append(myOrgs, membership.Namespace)
+					break
+				}
+			}
+		}
+		if len(myOrgs) == 1 {
+			err := setUserDefaultOrganization(ctx, r.Client, user.Name, myOrgs[0])
+			errGroup = multierr.Append(errGroup, err)
+		}
+	}
+
+	return ctrl.Result{}, errGroup
+}
+
+func setUserDefaultOrganization(ctx context.Context, c client.Client, userName string, orgName string) error {
+	user := controlv1.User{}
+	if err := c.Get(ctx, types.NamespacedName{Name: userName}, &user); err != nil {
```

Comment:
I might be mistaken but i think it's possible that the user does not exist yet. It's synced periodically from Keycloak. I think on a "not found" error we should create the user.

Reply:
I was hoping that using the organization membership's `status.resolvedUserRefs` list would sidestep that issue - since a user can only be resolved if it exists, right? I did assume though, I'll double check.

Reply:
I think resolved means "found in Keycloak". But my knowledge is pretty stale so double checking would be good!

Reply:
Checked again. The resolved user refs are set when a user's membership is updated in keycloak, which can only happen if the user exists (unless I'm misunderstanding something)

Reply:
> which can only happen if the user exists

in keycloak or in the control-api? The users are periodically synced from Keycloak to control-api. I'll bastel smth to check.

Reply:
yeah in keycloak. I'm more and more realizing that I'm making a bunch of assumptions on how this works...

Thanks for helping me check

Reply:
```
❯ k get  organizationmembers members -n swi-test -oyaml
apiVersion: appuio.io/v1
kind: OrganizationMembers
metadata:
  name: members
  namespace: swi-test
spec:
  userRefs:
  - name: sebastian.widmer.ac.test
status:
  resolvedUserRefs:
  - name: sebastian.widmer.ac.test
```

```
❯ ka get users.appuio.io sebastian.widmer.ac.test 
Error from server (NotFound): users.appuio.io "sebastian.widmer.ac.test" not found
```

Reply:
User does not need to exist in the control-api to be resolved. So let's just do a create on 404. Does not need any locking or anything since controller-runtime just retries on error.